lib/corefoundation/preferences.rb

```diff
@@ -23,49 +20,133 @@ module CF
+      CF.CFPreferencesAppSynchronize(application_id.to_cf)
+    end
 
-      def get!(key, application_id, username = nil, hostname = nil)
-        get(key, application_id, username, hostname) || (raise CF::Error, "Preference key `#{key}` not found")
-      end
+    # Calls the {#self.set} method and raise a `PreferenceSyncFailed` error if `false` is returned.
+    #
+    # @param [String] key Preference key to write
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @raise [PreferenceSyncFailed] If {#self.set} call returned false.
+    #
+    # @return [VALUE] Returns true if preference value is successfully written.
+    #
+    def self.set!(key, value, application_id, username = nil, hostname = nil)
+      hostname = arg_to_cf(hostname || ALL_HOSTS)
+      hostname = CF::Base.typecast(hostname).to_ruby
+      raise(CF::Exceptions::PreferenceSyncFailed.new(key, application_id, hostname)) unless set(key, value, application_id, username, hostname)
+    end
 
-      def set(key, value, application_id, username = CURRENT_USER, hostname = CURRENT_HOST)
-        CF.CFPreferencesSetValue(key.to_cf, value.to_cf, application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
-        CF.CFPreferencesAppSynchronize(application_id.to_cf)
-      end
+    # Checks whether a key exists in the preference domain. It'll also return false for an invalid domain.
+    #
+    # @param [String] key Preference key to read
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @return [TrueClass, FalseClass] Return true or false to indicate if it is a valid key.
+    #
+    def self.valid_key?(key, application_id, username = nil, hostname = nil)
+      domain_keys = list_keys(application_id, username, hostname)
+      domain_keys.include?(key)
+    end
 
-      def list_keys(application_id, username, hostname)
-        arr_ref = CF.CFPreferencesCopyKeyList(
-          application_id.to_cf,
-          arg_to_cf(username),
-          arg_to_cf(hostname)
-        )
-        CF::Array.new(arr_ref).to_ruby unless arr_ref.null?
-      end
+    private
 
-      private
+    # Get all existing keys for a preference domain.
+    #
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @return [Array<String>] Array of key names
+    #
+    # @visiblity private
+    def self.list_keys(application_id, username = nil, hostname = nil)
```

Comment:
If this method has no usage in here or `macos_userdefaults` resource, let's chuck this.

Reply:
Please refer to the `self.valid_keys?` method on your review above this where this is used.

Reply:
refer to the previous comment, https://github.com/chef/corefoundation/pull/11#r685152692

Reply:
closing since this is resolved